### ISSUE TITLE
exifr does not detect the tag "Projection Type" from a 360 camera file

### ISSUE THREAD

**Issue by avnermosh** [MikeKovarik/exifr#18]
I was not able to detect the tag "Projection Type" from a 360 camera file, which I downloaded from [here](https://www.orbix360.com/news/article/360-Image-Upload-Types)
I posted a comment to your comment in [here](https://stackoverflow.com/questions/59721220/can-exif-parser-extract-xmp-extensible-metadata-platform-tags
)
Also exifr has bigger sizes and slower download time when measured with https://bundlephobia.com/
![image](https://user-images.githubusercontent.com/26158466/75079535-ea057600-54bd-11ea-9f4e-619148fee0fb.png)

I guess by "efficient and blazing fast" you refer to the runtime after the package is already loaded?

Thanks

**Comment by MikeKovarik** [comment]
Hello, a lot of good questions, let me elaborate.

### xmp

That is correct. the property "Projection Type" is stored in XMP format, which is different to EXIF/TIFF which is the main focus of this library. Problem with XMP is that it's based on XML which i wildly complex. There's a note about it at the bottom of [advanced](https://github.com/MikeKovarik/exifr#advanced) in readme saying: "*Exifr does not come with an XML parser out of the box to keep the library simple and light-weight. There's plenty of XML parsers on npm. Exifr only extracts the XMP string and you can parse it.*"

There are two problems with parsing XML,
1) There isn't shared API that can be used both in node and in browser. Therefore exifr would need a dependency - [XMP parser](https://www.npmjs.com/search?q=xml%20parser) - at least for node.js, and these are pretty complex and range from 100kb up to 500kb. Even if i wanted to avoid that in brower by utilizing `window.DOMParser`, it would take a lot of code to stich both XMP solutions together.
2) Complexity of XML format. Specificaly an xml tag can have both attributes and children. This isn't easily translatable 1:1 to JSON and oppinions on how to do that vary.

**TLDR:** I didn't want this project to have a dependencies, let alone this large. And neither i wanted to force complex and ugly output (due to the children+attribute), or leave out some of the data because of it.

That being said **however**, over the weekend i worked on a proof-of-concept tiny XMP parser. If that turns out to work, I'd include it in the next release.

## bundle size

Also correct, but not quite.

Bundlephobia looks at the main entry point for the library, which is the full bundle. npm even states the size 588kB and 27 files. Let me explain.

Nowadays npm serves not only npm packages but also frontend ones. There's a big pressure to keep the things small in the browser with tree shaking and dead code removal, lazy loading and gzipping. Whereas in Node.js you don't need to worry about that.

Exifr contains a lot of parsers, because each metadata is stored in different format (TIFF is different from IPTC wich is different from ICC, and XMP, etc...). Each metadata segment is stored as enum values and a dictionary is needed to translate these numeric values. Dicts are especially large as they take up about half of the lib's size. If you add all that up. You get the 63KB. But sometimes you don't need all those parsers and features. So exifr v3 is modular. And you can pick just what you need.

Exifr comes in [four bundles](https://github.com/MikeKovarik/exifr#installation): core, mini, lite and full. Full (60kb) contains everything and is meant for use in Node.js. It is the default module (as stated in package.json), so bundlephobia picks that up and says 63kb.
Whereas in browsers the lite or mini are [recommended](https://github.com/MikeKovarik/exifr#installation). Those contain the bare minimum to read jpg files and get the most frequently used tags like resolution, orientation, exposure info, gps coords etc. These bundles are light and cover the most usecases people on the web have when it comes to exif. So I wanted to advertise that size for the browser users.

Plus, published to npm are not only the pre-built bundles in dist/, but also source code from src/. For when full/lite/mini bundles don't suit your needs. This brings the file count to 27 and module size to that 500+kb.
Yes we, the community, keep joking about the size of node_modules. But if I were to take the ultra modular approach of splitting exifr into a dozen little plugin modules, we'd end up with more accurate file size (for node people) but a lot more of readmes, licences and package.jsons, etc... I didn't want to this Babel or [turf v6](https://www.npmjs.com/search?q=%40turf) route. From my point of view it isn't worth it because it causes confusion, lot of lookup and npm installs. And majority of the modules get installed anyway, just to get the basic functions working.
It can scare of the first time users by asking "what do I need to get it to work" instead of just just getting it up and running fast.

**TLDR:** So exifr is modular and comes with everything, but in one large module, instead of dozen small ones.

Sidenote: Other libraries have dictionaries with just the basic tags like ISO, FNumer, etc... Whereas exifr also contains the less-frequent ones, or proprietary tags only used by certain companies. This means wider support but larger file size. It's up to every individual's opinion whether that's good or bad. But at least the less frequent tags are in separate file which is only part of the full bundle and not in the lite nor mini.

### performance

This is topic for another long discussion, but to keep it simple today: the speedup comes from letting exifr handle the [file reading](https://github.com/MikeKovarik/exifr#chunked-reader). Secondly, exifr is optimized around processing large ammount of files, hundreds of photos, gigabytes of data. As can be seen in the gif at the top of the readme.

If exifr is given already buffered binary data, there isn't much that can be done and speed can be comparable to other libraries. With one exception. If you are looking for a needle in the haystack - e.g. only want to get image's orientation, or GPS coords, etc... Exifr can utilize filter to only pick certain tags, or skip unwanted ones (like ignoring makernotes). And parsing stops once all wanted tags are found.

Again. If you use exifr to categorize whole galleries of files, this (along with the chunked reading) can be make the difference of milliseconds, rather than seconds to process.

PS: I could have included all of this in readme. But at first the readme was gigantic. It needed to be cut down for brevity and because I also took inspiration from other npm modules where simplicity seems to be the norm, whereas for example 'request' had me completely lost.

I hope this explains you questions :)

**Comment by MikeKovarik** [comment]
Hello @avnermosh, I hope the last message didn't scare you off.

**Status update:** I implemented the XMP parser over the last week. With it you'll be able to extract that "Projection Type" tag. It's merged in master and passes tests so you can try it out while I'm finishing bugfixes and other work. I hope to release 4.0.0 sometime this week.

Also the [homepage](http://mutiny.cz/exifr/) is deployed from master so you can drop your image and try it out.

**Comment by MikeKovarik** [comment]
Fixed In 4.0.0
